File: JavaScript/6js 1/findMissing.js
/*
|--------------------------------------------------------------------------
| Find the missing term in an Arithmetic Progression
|--------------------------------------------------------------------------
|
*/

function findMissing(list) {
   const n = list.length;
   const x1 = list[1] - list[0];
   const x2 = list[n - 1] - list[n - 2];
   const x = Math.abs(x1) > Math.abs(x2) ? x2 : x1;
   for (let i = 1; i < n; i++) {
      if (list[i] - list[i - 1] !== x) {
         return list[i - 1] + x;
      }
   }
}


function findMissing(list) {
   return (
      ((list[0] + list[list.length - 1]) / 2) * (list.length + 1) -
      list.reduce((pre, val) => pre + val)
   );
}


function findMissing(list) {
   let pattern = (list[list.length - 1] - list[0]) / list.length;
   let i = 0;
   while (list[i] + pattern === list[i + 1]) i++;
   return list[i] + pattern;
}


console.log(findMissing([1, 3, 4]), 2);

File: Python/6py 1/find_missing.py
"""
|--------------------------------------------------------------------------
| Find the missing term in an Arithmetic Progression
|--------------------------------------------------------------------------
|
"""

def find_missing(sequence):
   n = len(sequence)
   x1 = sequence[1] - sequence[0]
   x2 = sequence[n - 1] - sequence[n - 2]
   x = abs(x1) if abs(x1) < abs(x2) else x2
   for i in range(1, n):
      if sequence[i] - sequence[i - 1] != x:
         return sequence[i - 1] + x
      

def find_missing(sequence):
   return (sequence[-1] + sequence[0]) * (len(sequence) + 1) / 2 - sum(sequence)


def find_missing(sequence):
   interval = (sequence[-1] - sequence[0])/len(sequence)
   for previous, item in enumerate(sequence[1:]):
      if item - sequence[previous] != interval:
         return item - interval


print(find_missing([1, 2, 3, 4, 6, 7, 8, 9]), 5)
print(find_missing([1, 3, 4, 5, 6, 7, 8, 9]), 2)